hera_cal/abscal.py

```diff
@@ -0,0 +1,484 @@
+    been redundantly calibrated.
+
+    Be warned that the linearizations of the phase solvers suffer from phase wrapping
+    pathologies, meaning that a delay calibration should generally precede a phs_logcal
+    or a TT_phs_logcal solution.
+    """
+    def __init__(self, model, data, wgts=None, antpos=None, freqs=None, times=None, pols=None):
+        """
+        AbsCal object used to for phasing and scaling visibility data to an absolute reference model.
+
+        Parameters:
+        -----------
+        model : dict of visibility data of refence model, type=dictionary
+            keys are antenna pair + pol tuples, values are complex ndarray visibilities
+            these visibilities are 2D arrays with the [0] axis indexing time,
+            the [1] axis indexing frequency and (optionally) [2] axis indexing polarization
+
+        data : dict of visibility data of measurements, type=dictionary
+            keys are antenna pair + pol tuples (must match model), values are
+            complex ndarray visibilities, with shape matching model
+
+        wgts : dict of weights of data, type=dictionry, [default=None]
+            keys are antenna pair + pol tuples (must match model), values are real floats
+            matching shape of model and data
+
+        antpos : type=dictionary, dict of antenna position vectors in ENU frame in meters.
+                 origin of coordinates does not matter.
+                 keys are antenna integers and values are 2D or 3D ndarray
+                 position vectors in meters (topocentric coordinates),
+                 with [0] index containing X (E-W) distance, and [1] index Y (N-S) distance.
+                 Can be generated from a pyuvdata.UVData instance via
+                 ----
+                 #!/usr/bin/env python
+                 uvd = pyuvdata.UVData()
+                 uvd.read_miriad(<filename>)
+                 antenna_pos, ants = uvd.get_ENU_antpos()
+                 antpos = dict(zip(ants, antenna_pos))
+                 ----
+                 This is needed only for Tip Tilt phase calibration.
+
+        freqs : ndarray of frequency array, type=ndarray, dtype=float
+            1d array containing visibility frequencies in Hz. Needed to write gain solutions
+            out to calfits.
+    
+        times : ndarray of time array, type=ndarray, dtype=float
+            1d array containing visibility times in Julian Date. Needed to write out to calfits.
+
+        pols : ndarray of polarization array, type=ndarray, dtype=int
+            array containing polarization integers in pyuvdata.UVData.polarization_array 
+            format. Needed to write out to calfits.
+        """
+        # get shared keys
+        self.keys = sorted(set(model.keys()) & set(data.keys()))
+
+        # append attributes
+        self.model = model
+        self.data = data
+
+        # setup frequencies
+        self.freqs = freqs
+        if self.freqs is None:
```

Comment:
Looks like lines 110 through 113 are not used anywhere. Why do you need `self.Nfreqs`?